docs/api_requests/block-vhost-user.md

```diff
@@ -0,0 +1,189 @@
+handling of guests's requests. The UDS socket is only used for control purposes
+and does not participate in the dataplane.
+
+Firecracker only implements a vhost-user frontend. Users are free
+to choose from [existing open source backends](#backends) or implement their
+own.
+
+## Topology
+
+Each vhost-user device connects to its own UDS socket. There is no way for
+multiple devices to share a single socket, as there is no way to differentiate
+messages related to devices.
+
+Each device can be served by a separate backend or a single backend can serve
+multiple devices.
+
+## Interactions with the backend
+
+There are two points when the vhost-user frontend communicates with
+the backend:
+
+1. Device initialisation. When a vhost-user device is created, Firecracker
+   connects to the corresponding UDS socket and negotiates Virtio and Vhost
+   features with backend and retrieves device configuration.
+1. Device activation. When the guest driver finishes setting up the device,
+   Firecracker shares memory tables and Virtio queue information with the backend.
+   As a part of this, Firecracker shares file descriptors for guest's memory
+   regions, as well as file descriptors for queue notifications.
+
+## Advantages
+
+While vhost-user block is considered an optimisation to Firecracker IO, a naive
+implementation of the backend is not going to improve performance.
+
+The major advantage of using a vhost-user device is the backend can implement
+custom processing logic. It can use intelligent algorithms to serve block
+requests, eg by fetching the block device data over the network or using
+encryption or using sophisticated readahead logic. In such cases,
+the performance improvement will be coming from the fact that the custom logic
+is implemented in the same process that handles Virtio queues, which reduces
+the number of required context switches.
+
+## Backends
+
+There are a number of open source implementations of a vhost-user backend
+available for reference:
+
+1. [Qemu backend](https://github.com/qemu/qemu/tree/master/contrib/vhost-user-blk)
+1. [Cloud Hypervisor backend](https://github.com/cloud-hypervisor/cloud-hypervisor/tree/main/vhost_user_block)
+1. [crosvm backend](https://github.com/google/crosvm/blob/main/devices/src/virtio/vhost/user/device/block.rs)
+1. [SPDK backend](https://github.com/spdk/spdk/blob/master/lib/vhost/vhost_blk.c)
+
+## Security considerations
+
+### Guest memory sharing
+
+By design, a vhost-user frontend must share file descriptors
+of all guest memory regions to the backend. In order to achive that,
+guest memory is created as a [memfd](https://man7.org/linux/man-pages/man2/memfd_create.2.html)
+and mapped as `MAP_SHARED`.
+
+#### File descriptor in procfs
+
+An open `memfd` is reflected in `procfs` as any other open file descriptor:
+
+```shell
+$ ls -l /proc/{pid}/fd | grep memfd
+lrwx------ 1 1234 1234 64 Nov  2 13:39 32 -> /memfd:guest_mem (deleted)
```

Comment:
Is the `(deleted)` part correct here?

Reply:
If this is correct, feel free to resolve the comment.

Reply:
yes, this is correct. I don't know for sure what it means, but I think it may be related to the fact the file is not visible at filesystem level.